### PR TITLE
Enable hot-standby in utility mode

### PR DESCRIPTION
**Pull request by gfphoenix78**
~~To run pg_auto_failover for GPDB, we need some adapter code
for some reasons:~~
~~1. Enable hot-standby feature but only in utility mode.
        Calling cdb_setup() in hot-standby should be OK.
2. Fix `cdbcomponent_getComponentInfo()` issue.
        Currently, cdbcomponent_getComponentInfo() can't work well
        in hot-standby, it gets the coordinator's entry on the standby.
3. Fix ready status for hot-standby.
4. Set proper `dbid` and `content` for monitor instance.
        Monitor has dummy dbid and content. -1 is OK~~

Enable hot-standby in utility mode

1. When the standby turns on hot_standby, the ready status changed.
     pg_ctl can't properly wait for the ready status for the standby.

2. To connect to the GPDB cluster with multi-host/port in connection info,
     connecting to the hot-standby in dispatch(the default) mode is required,
     unless we change the retry policy in libpq. However, we're not intented
     to support distributed query in this patch. So, we force to use utility
     role when connecting to hot-standby.

syncrep is an issue, but it doesn't prevent from running pg_auto_failover.
We'll fix the issue in a later patch.

- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by gfphoenix78** [comment]
Note: `cdbcomponent_getComponentInfo()` is not completely fixed yet, please skip this PR now.

**Comment by ashwinstar** [comment]
> Note: `cdbcomponent_getComponentInfo()` is not completely fixed yet, please skip this PR now.

Feel free to close and open again later when ready

File: src/backend/cdb/cdbutil.c [hunk 1]
```diff
@@ -982,7 +982,15 @@ cdbcomponent_getComponentInfo(int contentId)
...
-		cdbInfo = &cdbs->entry_db_info[0];	
+		int i;
+		for (i = 0; i < cdbs->total_entry_dbs; i++)
+		{
+			if (cdbs->entry_db_info[i].config->dbid == GpIdentity.dbid)
+			{
+				cdbInfo = &cdbs->entry_db_info[i];	
+				break;
+			}
+		}
```
Comment (yaowangm) [inline src/backend/cdb/cdbutil.c:993]: It seems you adapted a loop to get cdbInfo in cdbs by comparing entry_db_info[i].config->dbid and GpIdentity.dbid. As my understanding, it's a more general approach because returning cdbs->entry_db_info[0] is incorrect for hot-standby case.

I am considering something about performance optimization. Is it feasible to return cdbs->entry_db_info[0] for not  hot-standby case, and perform the loop only for hot-standby? Like this:

```
if(!HotStandbyActive())
{
  return &cdbs->entry_db_info[0];	
}
else
{
  // perform the loop...
}
```

File: src/backend/postmaster/postmaster.c [hunk 2]
```diff
@@ -998,6 +998,15 @@ PostmasterMain(int argc, char *argv[])
 	if (!SelectConfigFiles(userDoption, progname))
 		ExitPostmaster(2);
 
+    elog(LOG, "role=%d, dbid=%d, segindex=%d\n", (int)Gp_role, GpIdentity.dbid, GpIdentity.segindex);
```
Comment (yaowangm) [inline src/backend/postmaster/postmaster.c:1001]: It seems the indent is incorrect.

File: src/bin/pg_ctl/pg_ctl.c [hunk 3]
```diff
@@ -584,13 +578,24 @@ wait_for_postmaster(pgpid_t pm_pid, bool do_checkpoint)
...
+	{
+		char		standby_file[MAXPGPATH];
+		int			fd;
+		snprintf(standby_file, sizeof(standby_file), "%s/standby.signal", pg_data);
+		fd = open(standby_file, O_RDONLY);
+		is_secondary = fd >= 0;
+		if (fd >= 0)
+			close(fd);
+	}
+
```
Comment (yaowangm) [inline src/bin/pg_ctl/pg_ctl.c:598]: 1. It seems the code is to detect if current node is secondary by opening standby_file, and only output is `is_secondary`. Could you add some comments to describe the code?
2. Is there any reason to put the code into a separated anonymous block? Is it better to use a separated inline function or something for code reuse?
3. In a big picture, I prefer it is not an ideal approach to determine `is_secondary` by opening standby_file because IO operation is always costly. Is there any more general approach to detect that? I suppose we should always know if it is secondary when the cluster starts. Is it correct?

File: src/bin/pg_ctl/pg_ctl.c [hunk 4]
```diff
@@ -631,8 +636,10 @@ wait_for_postmaster(pgpid_t pm_pid, bool do_checkpoint)
 				 */
 				char	   *pmstatus = optlines[LOCK_FILE_LINE_PM_STATUS - 1];
 
-				if (strcmp(pmstatus, gpdb_master_distributed_mode?PM_STATUS_DTM_RECOVERED:PM_STATUS_READY) == 0 ||
-					strcmp(pmstatus, PM_STATUS_STANDBY) == 0)
+				if (strcmp(pmstatus, PM_STATUS_DTM_RECOVERED) == 0 ||
+					strcmp(pmstatus, PM_STATUS_STANDBY) == 0 ||
+					(strcmp(pmstatus, PM_STATUS_READY) == 0 &&
+					 (!gpdb_master_distributed_mode || is_secondary)))
```
Comment (yaowangm) [inline src/bin/pg_ctl/pg_ctl.c:642]: It seems you added the condition:

`(!gpdb_master_distributed_mode || is_secondary)`

Could you please add some comments to describe the scenario?

File: src/backend/cdb/cdbutil.c [hunk 5]
```diff
@@ -982,7 +982,15 @@ cdbcomponent_getComponentInfo(int contentId)
...
 		return cdbInfo;
```
Comment (yaowangm) [inline src/backend/cdb/cdbutil.c:994]: Is it feasible to add some assertions before returning cdbInfo (not just here, but all return codes)? e.g. we can return NULL only for some specific cases, etc.

File: src/backend/cdb/cdbutil.c [hunk 6]
```diff
@@ -1082,7 +1092,7 @@ cdb_setup(void)
 	 */
 	if (!IsBackgroundWorker &&
 		Gp_role == GP_ROLE_DISPATCH &&
-		!*shmDtmStarted)
+		(!*shmDtmStarted && !HotStandbyActive()))
```
Comment (yaowangm) [inline src/backend/cdb/cdbutil.c:1095]: Could you please add some comments to describe the logic?

File: src/backend/cdb/cdbutil.c [hunk 7]
```diff
@@ -982,7 +982,15 @@ cdbcomponent_getComponentInfo(int contentId)
 	/* entry db */
 	if (contentId == -1)
 	{
-		cdbInfo = &cdbs->entry_db_info[0];	
+		int i;
+		for (i = 0; i < cdbs->total_entry_dbs; i++)
```
Comment (asimrp) [inline src/backend/cdb/cdbutil.c:986]: This change won't be needed if we get rid of dbids wholesale.  It doesn't feel right to create contraptions around dbid, which is a useless entity for coordinator autofailover.  Trying to get rid of it seems more robust alternative.

File: src/backend/cdb/cdbutil.c (hunk 6 above)
Comment (asimrp) [reply src/backend/cdb/cdbutil.c:1095]: +1.  For example, we can add something like the following:
> In order to facilitate coordinator / standby autofailover, the standby needs to be able to run catalog queries.  Some third parties such as pg_auto_failover need to do so.  Full capability to execute distributed transactions is not needed.  Therefore, let such connections move forward even when DTM is initialized.

File: src/backend/postmaster/postmaster.c (hunk 2 above)
Comment (asimrp) [reply src/backend/postmaster/postmaster.c:1001]: Looks like a remnant of debugging, do you want to remove this change?

File: src/backend/postmaster/postmaster.c [hunk 8]
```diff
@@ -998,6 +998,15 @@ PostmasterMain(int argc, char *argv[])
...
+	if (Gp_role == GP_ROLE_UTILITY)
```
Comment (asimrp) [inline src/backend/postmaster/postmaster.c:1002]: Ooo, we definitely need some comments to explain this hack!

File: src/backend/replication/syncrep.c [hunk 9]
```diff
@@ -489,7 +489,6 @@ SyncRepInitConfig(void)
 	 */
 	if (IS_QUERY_DISPATCHER() && MyWalSnd->is_for_gp_walreceiver)
 	{
-		Assert(priority == 0);
```
Comment (asimrp) [inline src/backend/replication/syncrep.c:492]: Didn't we talk about removing this Greenplum-specific behavior specifically for coordinator autofailover?

File: src/bin/pg_ctl/pg_ctl.c (hunk 4 above)
Comment (asimrp) [reply src/bin/pg_ctl/pg_ctl.c:642]: Have you looked into how upstream deals with hot-standby in `pg_ctl`?  Now that we have hot-standby enabled, can we reuse the upstream logic for this purpose?

**Review by asimrp — COMMENTED** [review]
> Set proper `dbid` and `content` for monitor instance.
    Monitor has dummy dbid and content. -1 is OK

This part doesn't feel right.  Our goal should be to avoid any changes to `pg_auto_failover` that are related to dbid.  Making `pg_auto_failover` aware of dbids is not a good idea.

File: src/backend/replication/syncrep.c (hunk 9 above)
Comment (gfphoenix78) [reply src/backend/replication/syncrep.c:492]: Keeping the assert will cause the database to fail when enabled pg_auto_failover.
The Greenplum-specific behavior in replication will be removed in a later patch, which affects several tests.
This PR will do minimal work to run GPDB with pg_auto_failover.

**Comment by gfphoenix78** [comment]
pg_ctl checks the status in postmaster.pid. In postgres, it only checks whether it's `ready` or `standby`.
Here are ready-status table for both postgres and gpdb.

In postgres:
| Role | Hot-Standby | Ready Status |
| :---: | :--: | :--: |
|primary|on|ready|
|secondary|on|ready|
|primary|off|ready|
|secondary|off|standby|

In gpdb:
| Role | Hot-Standby | Ready Status |
| :---: | :--: | :--: |
|coordinator|on|dtmready|
|standby|on|ready|
|primary|on|ready|
|mirror|on|ready|
|coordinator|off|dtmready|
|standby|off|standby|
|primary|off|ready|
|mirror|off|standby|

**Comment by gfphoenix78** [comment]
When the status is `ready`, the PG instance is ready in GPDB if it's not the coordinator(i.e. not dispatch or it's standby).

File: src/bin/pg_ctl/pg_ctl.c (hunk 3 above)
Comment (gfphoenix78) [reply src/bin/pg_ctl/pg_ctl.c:598]: It's no performance issue. It's called in pg_ctl and the function will wait for the postmaster to be ready.
The startup process also detects the existence of the file `standby.signal` to run as standby.

File: src/bin/pg_ctl/pg_ctl.c (hunk 4 above)
Comment (gfphoenix78) [reply src/bin/pg_ctl/pg_ctl.c:642]: `ready` is READY in Postgres, but it's not really READY in coordinator. See the below ready-status tables.

File: src/backend/postmaster/postmaster.c (hunk 2 above)
Comment (yaowangm) [reply src/backend/postmaster/postmaster.c:1001]: As discussed, the code is for monitor only. For monitor, it is possible that GpIdentity.dbid and  		GpIdentity.segindex are not set at the time, so they might be less than 0. For the case, we need to set they to -1 for being recognized.

Hao will add relevant comments.

**Review by yaowangm — COMMENTED** [review]
For removing dbids wholesale, generally I agree it's a correct direction, but we need to think a little bit more about the details and steps. We can start with avoiding dbids in new code as much as possible, then removing all existing dbids in a big delivery with detailed tests. Maybe we can have another discussion.

File: src/bin/pg_ctl/pg_ctl.c [hunk 10]
```diff
@@ -564,6 +559,17 @@ start_postmaster(void)
...
+static bool
+is_secondary_instance(const char *pg_data)
+{
+	char		standby_file[MAXPGPATH];
+	int			fd;
+	snprintf(standby_file, sizeof(standby_file), "%s/standby.signal", pg_data);
+	fd = open(standby_file, O_RDONLY);
+	if (fd >= 0)
+		close(fd);
+	return fd >= 0;
```
Comment (paul-guo-) [inline src/bin/pg_ctl/pg_ctl.c:571]: `access()` is better than `open()+close()`

File: src/bin/pg_ctl/pg_ctl.c [hunk 11]
```diff
@@ -169,15 +169,10 @@ static void free_readfile(char **optlines);
...
+static bool is_secondary_instance(const char *pg_data);
 static WaitPMResult wait_for_postmaster(pgpid_t pm_pid, bool do_checkpoint);
 static bool postmaster_is_alive(pid_t pid);
 
-static char postopts_file[MAXPGPATH];
-static char backup_file[MAXPGPATH];
-static char promote_file[MAXPGPATH];
-static char pid_file[MAXPGPATH];
-static char backup_file[MAXPGPATH];
-
```
Comment (paul-guo-) [inline src/bin/pg_ctl/pg_ctl.c:180]: Why remove these files? I assume this change is not intended?

File: src/backend/storage/lmgr/proc.c [hunk 12]
```diff
@@ -345,9 +345,13 @@ InitProcess(void)
...
 	 * such as mppSessionId being valid and mppIsWriter set to true.
 	 */
-	if (am_walsender || am_ftshandler ||
-		IsFaultHandler)
+	if (am_walsender || am_ftshandler || IsFaultHandler ||
+		(GpIdentity.segindex == -1 && HotStandbyActive()))
+	{
 		Gp_role = GP_ROLE_UTILITY;
+		if (GpIdentity.segindex == -1 && HotStandbyActive())
+			elog(WARNING, "force to utility mode in hot-standby");
```
Comment (paul-guo-) [inline src/backend/storage/lmgr/proc.c:353]: "Force to run in utility mode in hot standby"?

Please use ereport()

File: src/bin/pg_ctl/pg_ctl.c [hunk 13]
```diff
@@ -583,13 +589,11 @@ static WaitPMResult
...
+	bool		is_coordinator;
 
-	/* check if starting GPDB master in distributed mode */
-	if (strstr(post_opts, "gp_role=dispatch"))
-		gpdb_master_distributed_mode = true;
-	else
-		gpdb_master_distributed_mode = false;
+	/* check if starting GPDB coordinator in distributed mode */
+	is_coordinator = strstr(post_opts, "gp_role=dispatch") != NULL
+                        && !is_secondary_instance(pg_data);
```
Comment (paul-guo-) [inline src/bin/pg_ctl/pg_ctl.c:596]: Actually maybe we  remove`gpdb_master_distributed_mode` or ` is_coordinator`, instead use the code below later to determine if postmaster is "ready" and this just introduces one line of difference against upstream code.
```
if (strcmp(pmstatus, PM_STATUS_READY) == 0 ||
strcmp(pmstatus, PM_STATUS_DTM_RECOVERED) == 0 ||
                      strcmp(pmstatus, PM_STATUS_STANDBY) == 0)
```

If you really want to judge more strictly, you need to write with more logic with different kind of nodes.

File: src/backend/storage/lmgr/proc.c (hunk 12 above)
Comment (gfphoenix78) [reply src/backend/storage/lmgr/proc.c:353]: elog(...) is used in the next few lines and it looks more consistent.

File: src/bin/pg_ctl/pg_ctl.c (hunk 13 above)
Comment (gfphoenix78) [reply src/bin/pg_ctl/pg_ctl.c:599]: It's no a big deal. We'll get some annoying FATAL messages, or pg_ctl/gpstart returns OK, but the dtx recovery may fail, that's what we need to avoid.

**Comment by gfphoenix78** [comment]
When connecting to the hot standby in `dispatch` mode, the gp_role is forcibly set to `utility`, we needn't change the code in `cdbcomponent_getComponentInfo()` and `cdb_setup()`.

File: src/backend/storage/lmgr/proc.c (hunk 12 above)
Comment (paul-guo-) [reply src/backend/storage/lmgr/proc.c]: elog() is not encouraged in upstream. We should not use that in new code.

File: src/bin/pg_ctl/pg_ctl.c (hunk 11 above)
Comment (gfphoenix78) [reply src/bin/pg_ctl/pg_ctl.c:180]: They are duplicated.

Comment (zhangh43) [reply src/bin/pg_ctl/pg_ctl.c:180]: see line 106

File: src/backend/storage/lmgr/proc.c [hunk 14]
```diff
@@ -344,10 +344,17 @@ InitProcess(void)
...
+	 * RecoveryInProgress() to see if we are in hot standby, because
+	 * HotStandbyActive() is still true after promotion.
 	 */
-	if (am_walsender || am_ftshandler ||
-		IsFaultHandler)
+	if (am_walsender || am_ftshandler || IsFaultHandler ||
+		(GpIdentity.segindex == -1 && RecoveryInProgress()))
+	{
 		Gp_role = GP_ROLE_UTILITY;
+		if (GpIdentity.segindex == -1 && RecoveryInProgress())
```
Comment (zhangh43) [inline src/backend/storage/lmgr/proc.c:355]: we don't need to set `Gp_role = GP_ROLE_UTILITY` in this PR.
Support dispatch mode hot standby should be handled in another PR.

File: src/bin/pg_ctl/pg_ctl.c [hunk 15]
```diff
@@ -631,7 +638,11 @@ wait_for_postmaster(pgpid_t pm_pid, bool do_checkpoint)
 				 */
 				char	   *pmstatus = optlines[LOCK_FILE_LINE_PM_STATUS - 1];
 
-				if (strcmp(pmstatus, gpdb_master_distributed_mode?PM_STATUS_DTM_RECOVERED:PM_STATUS_READY) == 0 ||
+				/*
+				 * The READY status for coordinator is `dtmready`, while the READY
+				 * status is really ready for other nodes.
```
Comment (zhangh43) [inline src/bin/pg_ctl/pg_ctl.c:643]: READY for coordinator in utility mode is 'PM_STATUS_READY', READY for non hot standby is 'PM_STATUS_STANDBY', READY for hot standby is `PM_STATUS_READY `.

**Review by zhangh43 — APPROVED** [review]
Note that the postmaster READY status for hot standby is ''PM_STATUS_READY' which is not handled by the previous pg_ctl logic. This PR could avoid the pg_ctl hung issue in `wait_for_postmaster`. Upstream doesn't have this issue, since it doesn't have status `PM_STATUS_DTM_RECOVERED`

Others looks good to me.

File: src/backend/storage/lmgr/proc.c (hunk 14 above)
Comment (gfphoenix78) [reply src/backend/storage/lmgr/proc.c:355]: Yes, we currently only need to connect to the hot-standby, and doesn't support dispatching queries.

File: src/bin/pg_ctl/pg_ctl.c (hunk 15 above)
Comment (gfphoenix78) [reply src/bin/pg_ctl/pg_ctl.c:643]: Yes, in utility mode, the READY status for the instance is the same as the upstream. See the above tables.

File: src/backend/storage/lmgr/proc.c (hunk 14 above)
Comment (jimmyyih) [inline src/backend/storage/lmgr/proc.c:355]: Why do we need to enforce the standby coordinator to be in `GP_ROLE_UTILITY`?  Is it to prevent users from connecting to the standby coordinator in hot standby mode when pg_auto_failover is running (to prevent query dispatches from the standby coordinator to the primary segments)?

If so, I think it would be better to modify `gpinitstandby` instead to not start up the standby coordinator with `-c gp_role=dispatch` in its `pg_ctl` call (this would make it default to use gp_role utility).

However, my suggestion or this PR's change might have an oversight.  What would happen when the standby coordinator is promoted.  Would the `Gp_role` stay as utility or would it change to dispatch?

Comment (gfphoenix78) [reply src/backend/storage/lmgr/proc.c:355]: > Why do we need to enforce the standby coordinator to be in GP_ROLE_UTILITY? Is it to prevent users from connecting to the standby coordinator in hot standby mode when pg_auto_failover is running (to prevent query dispatches from the standby coordinator to the primary segments)?

Yes, it prevents the standby coordinator to dispatch queries. To support multiple-host/port in the connection info in auto-failover, the client is expected to connect to the standby successfully. But we have some other issues to support dispatching queries on the standby coordinator.

> If so, I think it would be better to modify gpinitstandby instead to not start up the standby coordinator with -c gp_role=dispatch in its pg_ctl call (this would make it default to use gp_role utility).

It breaks the promise for coordinator auto-failover. If the standby coordinator was started in utility mode, it can't dispatch queries even after promotion. The gp_role is unchanged after promotion.

The best patch is to fully enable hot-standby including dispatching queries. I've not many spikes about it. Does your patch work well for dispatching queries in the hot-standby coordinator?

**Review by jimmyyih — APPROVED** [review]
There is still concern about how the forcing of utility mode on a coordinator (content id -1) in hot standby mode can be restricting (e.g. when we need to actually do a dispatch and the postmaster was specifically brought up with -c gp_role=dispatch).  It's okay for now I guess to merge this in and we can work on that concern in a later PR as mentioned in various comments in this PR.

File: src/backend/storage/lmgr/proc.c (hunk 14 above)
Comment (ashwinstar) [reply src/backend/storage/lmgr/proc.c:355]: > 
> The best patch is to fully enable hot-standby including dispatching queries. I've not many spikes about it. Does your patch work well for dispatching queries in the hot-standby coordinator?

That's big project (support multi-coordinators for read-only access), need to take care of correctly creating distributed snapshots and locking and other things. So, lets not take that up now.

**Comment by gfphoenix78** [comment]
Thank you, all guys.
